scripts/kiwix-resources

```diff
@@ -0,0 +1,126 @@
+#!/usr/bin/env python3
+
+'''
+Copyright 2022 Veloman Yunkan <veloman.yunkan@gmail.com>
+
+This program is free software; you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation; either version 3 of the License, or any
+later version.
+
+This program is distributed in the hope that it will be useful, but
+WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program; if not, write to the Free Software
+Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+02110-1301, USA.
+'''
+
+import argparse
+import hashlib
+import os.path
+import re
+
+def read_resource_file(resource_file_path):
+    with open(resource_file_path, 'r') as f:
+        return [line.strip() for line in f]
+
+def list_resources(resource_file_path):
+    for resource_path in read_resource_file(resource_file_path):
+        print(resource_path)
+
+def get_resource_revision(base_dir, resource_path):
+    with open(os.path.join(base_dir, resource_path), 'rb') as f:
+        return hashlib.sha1(f.read()).hexdigest()[:8]
+
+resource_revisions = {}
+
+def fill_resource_revisions(resource_file_path):
+    base_dir = os.path.dirname(os.path.realpath(resource_file_path))
+    for resource in read_resource_file(resource_file_path):
+        resource_revisions[resource] = get_resource_revision(base_dir, resource)
+
+RESOURCE_WITH_CACHEID_URL_PATTERN=r'"((.*)/skin/([^"?]+))\?KIWIXCACHEID([^"]*)"'
+
+def set_cacheid(resource_matchobj):
+    path = resource_matchobj.group(1)
+    resource = 'skin/' + resource_matchobj.group(3)
+    extra_query = resource_matchobj.group(4)
+    cacheid = 'cacheid=' + resource_revisions[resource]
+    return '"' + path + '?' + cacheid + extra_query + '"'
+
+def preprocess_line(line):
+    if 'KIWIXCACHEID' in line:
+        line = re.sub(RESOURCE_WITH_CACHEID_URL_PATTERN, set_cacheid, line)
+        assert not 'KIWIXCACHEID' in line
+    return line
+
+def get_preprocessed_resource(srcpath):
+    modified_line_count = 0
+    preprocessed_lines = []
+    try:
+        with open(srcpath, 'r') as source:
+            for line in source:
+                ppline = preprocess_line(line)
+                if ppline != line:
+                    modified_line_count += 1
+                preprocessed_lines.append(ppline)
+        return "".join(preprocessed_lines), modified_line_count
```

Comment:
Why process each line separately ?
It should be possible to pass all the content to `re.sub`(and `re.subn` returns the number of occurrences replaced)

Reply:
Done in a very slightly different way. But I doubt that the result is a net optimization (if that was the point of this suggestion).

Reply:
It was not a optimization. More a simplification of the code. This code is not really a critical path, I prefer to have simpler code